library/std/src/thread/spawnhook.rs

```diff
@@ -0,0 +1,92 @@
+use crate::io;
+use crate::sync::RwLock;
+use crate::thread::Thread;
+
+static SPAWN_HOOKS: RwLock<
+    Vec<&'static (dyn Fn(&Thread) -> io::Result<Box<dyn FnOnce() + Send>> + Sync)>,
+> = RwLock::new(Vec::new());
```

Comment:
An alternative is to use a lock free linked list:

```rust
static SPAWN_HOOK: AtomicPtr<SpawnHook> = AtomicPtr::new(ptr::null_mut());

struct SpawnHook {
    next: Option<&'static SpawnHook>,
    hook: &'static (dyn Fn(&Thread) -> io::Result<Box<dyn FnOnce() + Send>> + Sync),
}

pub fn add_spawn_hook<F, G>(hook: F)
where
    F: 'static + Sync + Fn(&Thread) -> io::Result<G>,
    G: 'static + Send + FnOnce(),
{
    let mut next = SPAWN_HOOK.load(Ordering::Relaxed);
    let mut hook = Box::leak(Box::new(SpawnHook {
        next: unsafe { next.as_ref() },
        hook: Box::leak(Box::new(move |thread: &Thread| {
            let f: Box<dyn FnOnce() + Send> = Box::new(hook(thread)?);
            Ok(f)
        })),
    }));
    while let Err(n) =
        SPAWN_HOOK.compare_exchange_weak(next, hook, Ordering::Release, Ordering::Relaxed)
    {
        next = n;
        hook.next = unsafe { next.as_ref() };
    }
}
```

But that uses two Boxes per hook (one for the `dyn Fn` and one for the linked list node (`SpawnHook`). We could merge them into a single `Box` by storing the `Fn` by value in a `SpawnHook<F>`, but that results in a lot more complexity and much more subtle unsafe code, which is probably not worth it.